### PR TITLE
[CUDA] Drop CUDA 10 support

### PR DESCRIPTION
**Pull request by eqy**
CC @ptrblck @ngimel @malfet

### CODE REVIEW

**Review by malfet — APPROVED** [review]

**Comment by yaox12** [comment]
There's some legacy code defined under the `__CUDACC_VER_MAJOR__` macro, should we remove them too? For example,
https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/caffe2/utils/math_gpu.cu#L725

File: aten/src/ATen/Dispatch.h [hunk 1]
```diff
@@ -79,11 +79,7 @@ TORCH_API void record_kernel_function_dtype(std::string name);
 // Workaround for C10_UNUSED because CUDA 10.2 and below fails to handle unused
```
Comment (ngimel) [inline aten/src/ATen/Dispatch.h:79]: remove this comment also

File: c10/util/complex.h [hunk 2]
```diff
@@ -515,14 +515,7 @@ namespace c10_internal {
 template <typename T>
 C10_HOST_DEVICE constexpr thrust::complex<T>
 cuda101bug_cast_c10_complex_to_thrust_complex(const c10::complex<T>& x) {
-#if defined(CUDA_VERSION) && (CUDA_VERSION < 10020)
```
Comment (ngimel) [inline c10/util/complex.h:518]: similarly here, `cuda101bug...` uses should just be replaced with  static_cast, as the comment suggests

**Comment by eqy** [comment]
Hmm, what macro should be used instead of `USE_ROCM` or `TORCH_HIP_VERSION` in `c10/util/*`? The former isn't defined there for hip builds and the latter seems to still break on  the test CI runs (e.g., https://github.com/pytorch/pytorch/actions/runs/3625677594/jobs/6114427598#logs )

**Comment by eqy** [comment]
CC @jithunnair-amd who might know more about the last issue

**Comment by eqy** [comment]
I've left the version guards in the `c10/cuda` files in for now as no one from AMD/ROCM has responded. Dropping the [WIP] tag.

File: aten/src/ATen/Dispatch.h [hunk 3]
```diff
@@ -79,11 +79,7 @@ TORCH_API void record_kernel_function_dtype(std::string name);
...
 // attribute in the type aliasing context. Keep name long and verbose to avoid
 // macro collisions.
-#if defined(__CUDACC__) && CUDA_VERSION < 11000
-#define C10_UNUSED_DISPATCH_CUDA_WORKAROUND
-#else
 #define C10_UNUSED_DISPATCH_CUDA_WORKAROUND C10_UNUSED
```
[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/Dispatch.h:79]: Nit: can we just replace `C10_UNUSED_DISPATCH_CUDA_WORKAROUND` with `C10_UNUSED` everywhere?

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**